config/v1/types_cluster_version.go

```diff
@@ -290,6 +305,93 @@ type Release struct {
 
+// ConditionalUpdate represents an update which is recommended to some
+// clusters on the version the current cluster is reconciling, but which
+// may not be recommended for the current cluster.
+// +k8s:deepcopy-gen=true
+type ConditionalUpdate struct {
+	// release is the target of the update.
+	// +required
+	Release Release `json:"release"`
+
+	// risks represents the range of issues associated with
+	// updating to the target release. The cluster-version
+	// operator will evaluate all entries, and only recommend the
+	// update if there is at least one entry and all entries
+	// recommend the update.
+	// +optional
+	Risks []ConditionalUpdateRisk `json:"risks,omitempty"`
+
+	// conditions represents the observations of the conditional update's
+	// current status. Known types are:
+	// * Evaluating, for whether the cluster-version operator will attempt to evaluate any risks[].matchingRules.
+	// * Recommended, for whether the update is recommended for the current cluster.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+}
+
+// ConditionalUpdateRisk represents a reason and cluster-state
+// for not recommending a conditional update.
+// +k8s:deepcopy-gen=true
+type ConditionalUpdateRisk struct {
+	// url contains information about this risk.
+	// +optional
+	URL URL `json:"url,omitempty"`
+
+	// name is the CamelCase reason for not recommending a
+	// conditional update, in the event that matchingRules match the
+	// cluster state.
+	// +optional
+	Name string `json:"name,omitempty"`
+
+	// message provides additional information about the risk of
+	// updating, in the event that matchingRules match the cluster
+	// state. This is only to be consumed by humans. It may
+	// contain Line Feed characters (U+000A), which should be
+	// rendered as new lines.
+	// +optional
+	Message string `json:"message,omitempty"`
+
+	// matchingRules is a slice of conditions for deciding which
+	// clusters match the risk and which do not. The slice is
+	// ordered by decreasing precedence. The cluster-version
+	// operator will walk the slice in order, and stop after the
+	// first it can successfully evaluate. If no condition can be
+	// successfully evaluated, the update will not be recommended.
+	// +optional
+	MatchingRules []ClusterCondition `json:"matchingRules,omitempty"`
+}
+
+// ClusterCondition is a union of typed cluster conditions.  The 'type'
+// property determines which of the type-specific properties are relevant.
+// When evaluated on a cluster, the condition may match, not match, or
+// fail to evaluate.
+// +k8s:deepcopy-gen=true
+type ClusterCondition struct {
+	// type represents the cluster-condition type. This defines
+	// the members and semantics of any additional properties.
+	// +required
```

Comment:
Set it up like an enum and tag it so you get the vallidation for future us.

Reply:
I briefly pushed f95ef150 -> 7af2373d adding `+kubebuilder:validation:Enum={"always","promql"}` here, but have since gone back to f95ef150.  Because of the choice the enhancement made with [discriminating unions][1] we will not be able to store any metadata associated with unrecognized type.  But we do need to at least record the fact that we saw an unrecognized type, so we can include "And I don't know how to evaluate `$UNRECOGNIZED_TYPE` for `$RISK`" in our reasons for going `Recommended=Unknown` for that update target.  I can put the `kubebuilder:validation:Enum` back in if you're sure you would prefer the CVO to silently filter out any unrecognized cluster-condition types without complaining about them anywhere user-visible.  Thoughts?

[1]: https://github.com/openshift/enhancements/blob/e539d55691f2710d4fae0a34ac7c4b6b4e87fa10/enhancements/update/targeted-update-edge-blocking.md#discriminating-union-implementation

Reply:
Our past experience has been that an explicit enumeration makes it easier to keep the data clean in spite of some difficulties in rollout.

Reply:
I love clean data.  But enum or not seems orthogonal to that here, because the CVO is very unlikely to be inventing its own `type` values.  Instead, it is just passing along values that the `upstream` update service supplied in Cincinnati JSON.  So if we get a `type` value we recognize, great, in it goes.  If we get a `type` value we don't recognize, then what?

 * In enum-land, we won't be able to push it out to the ClusterVersion, because the API server would reject the not-in-the-enum type.  So we just silently drop it.  Nobody downstream of the CVO ever knows that the `upstream` update service suggested the new-type rule.
* In no-enum-land, we can include it (although because we picked [a discriminating union implementation that does not support external extension][1], we cannot include any type-specific configuration).  And then we can complain about it like "And I don't know how to evaluate `$UNRECOGNIZED_TYPE` for `$RISK`".  So folks downstream of the CVO will know that the `upstream` update service suggested the new-type rule.

That second option sounds better to me (I like transparency).  And this is not a user-writable field.  Even if a cluster admin maliciously injects some bogus `type` value, the CVO will come around and stomp it back to whatever the `upstream` was saying on its next sync.  So can you provide more details on how an enum here would make it easier to keep the data clean?

[1]: https://github.com/openshift/enhancements/blob/e539d55691f2710d4fae0a34ac7c4b6b4e87fa10/enhancements/update/targeted-update-edge-blocking.md#discriminating-union-implementation

Reply:
Ok, I'm still not on board with this, but I'm not the API approver.  a93e6749 -> 3b6de503 adds `+kubebuilder:validation:Enum={"always","promql"}`.